tests/src/Kernel/ConfigEntityTest.php

```diff
@@ -66,6 +76,46 @@ public function testConfigAction() {
   }
 
   /**
+   * Test saving a deeply nested rule.
+   *
+   * @todo This test currently fails if $strictConfigSchema = TRUE.
+   */
+  public function testNestedRuleSave() {
+    // This test is specific to reaction rules, so get the appropriate
+    // entity storage:
+    $storage = $this->container->get('entity.manager')->getStorage('rules_reaction_rule');
+    // Create a nested expression.
+    $and = $this->expressionManager->createAnd()
+      ->addCondition('rules_test_false')
+      ->addCondition('rules_test_true', ContextConfig::create()->negateResult())
+      ->negate();
+
+    $or = $this->expressionManager->createOr()
+      ->addCondition('rules_test_true', ContextConfig::create()->negateResult())
+      ->addCondition('rules_test_false')
+      ->addCondition($and);
```

Comment:
Scalar type hinting is still not supported in PHP 5.x (I'm not sure about PHP 7), but the following would catch this kind of error in a way that's much easier to debug:

      /**
       * {@inheritdoc}
       */
      public function createCondition($id) {
        if (!is_string($id)) {
          throw new \InvalidArgumentException;
        }
        return $this->createInstance('rules_condition', [
          'condition_id' => $id,
        ]);
      }

This throws an exception at the point of the error:

````
There was 1 error:

1) Drupal\Tests\rules\Kernel\ConfigEntityTest::testNestedRuleSave
InvalidArgumentException:

/var/www/drupal/modules/custom/rules/src/Engine/ExpressionManager.php:66
/var/www/drupal/modules/custom/rules/src/Engine/ConditionExpressionContainer.php:88
/var/www/drupal/modules/custom/rules/tests/src/Kernel/ConfigEntityTest.php:96
````

What do you think about doing this to prevent this kind of thing?  The code in the test is based on stuff that was in the tree back in September, so these are easy kinds of errors to make. Doing this would make these kinds of errors easier to debug.